app/src/org/commcare/android/database/app/models/UserKeyRecord.java

```diff
@@ -208,6 +267,56 @@ public boolean isPasswordValid(String password) {
                 hash.equals(UserKeyRecord.generatePwdHash(password, UserKeyRecord.extractSalt(hash))));
     }
 
+    public void assignPinToRecord(String pin, String password) {
+        this.passwordWrappedByPin = wrapPasswordWithPin(password, pin);
+    }
+
+    public byte[] getWrappedPassword() {
+        return passwordWrappedByPin;
+    }
+
+    public boolean hasPinSet() {
+        return passwordWrappedByPin.length > 0;
+    }
+
+    public boolean isPinValid(String pin) {
+        // Unwrap wrapped password with the PIN, and then check if the resulting password is correct
+        return isPasswordValid(getUnhashedPasswordViaPin(pin));
+    }
+
+    /**
+     * Returns the un-hashed password that was wrapped by the given PIN, or null if the given PIN
+     * is not valid to unwrap the wrapped password
+     */
+    public String getUnhashedPasswordViaPin(String pin) {
+        return unwrapPasswordWithPin(this.passwordWrappedByPin, pin);
+    }
+
+    /**
+     * Must be called with 1 of the 2 values set to null, which indicates that we should check
+     * based upon the other
+     */
+    public boolean isPasswordOrPinValid(String password, String pin) {
+        if (pin != null) {
+            return isPinValid(pin);
+        } else {
+            return isPasswordValid(password);
+        }
+    }
+
+    public static byte[] wrapPasswordWithPin(String password, String pin) {
+        return CryptUtil.wrapByteArrayWithString(password.getBytes(), pin);
+    }
```

Comment:
This can be `private`. It is only called once, I would remove it and just do `passwordWrappedByPin = CryptUtil.wrapByteArrayWithString(password.getBytes(), pin);` in `assignPinToRecord`. The name of the var you are assigning the result of `CryptUtil.wrapByteArrayWithString` to gives enough context that you don't need to create a helper method